lib/Reader/readerir.cpp

```diff
@@ -3240,232 +3272,171 @@ LandingPadInst *GenIR::createLandingPad(EHRegion *TryRegion,
+    static_assert(sizeof(mdToken) == sizeof(int32_t), "Unexpected token size");
+    IntegerType *TokenType = Type::getInt32Ty(LLVMContext);
+    Constant *TokenConstant = ConstantInt::get(TokenType, ClassToken);
+    BasicBlock *PadBlock = createPointBlock(Handler->StartMsilOffset, "Catch");
+    BasicBlock *ThunkBlock =
+        createPointBlock(Handler->StartMsilOffset, "Catch.Exn");
+    LLVMBuilder->SetInsertPoint(PadBlock);
+    Pad = Handler->CatchPad = LLVMBuilder->CreateCatchPad(
+        ThunkBlock, NextPadBlock, {TokenConstant}, "CatchPad");
+    LLVMBuilder->SetInsertPoint(ThunkBlock);
+    Function *GetException =
+        Intrinsic::getDeclaration(JitContext->CurrentModule,
+                                  Intrinsic::eh_exceptionpointer, {CaughtType});
+    Exception = LLVMBuilder->CreateCall(GetException, {Pad}, "exn");
+    LLVMBuilder->CreateBr(HandlerCodeBlock);
+    break;
+  }
+  case ReaderBaseNS::RegionKind::RGN_Finally: {
+    BasicBlock *PadBlock =
+        createPointBlock(Handler->StartMsilOffset, "Finally");
+    LLVMBuilder->SetInsertPoint(PadBlock);
+    Pad = Handler->CleanupPad = LLVMBuilder->CreateCleanupPad({}, "FinallyPad");
+    LLVMBuilder->CreateBr(HandlerCodeBlock);
+    // The in-flight exception is not available to a finally handler.
+    Exception = nullptr;
+    break;
+  }
+  case ReaderBaseNS::RegionKind::RGN_Fault: {
+    BasicBlock *PadBlock = createPointBlock(Handler->StartMsilOffset, "Fault");
+    LLVMBuilder->SetInsertPoint(PadBlock);
+    // Add an i32 argument to faults, to distinguish them from filters.
+    Value *Undef = UndefValue::get(Type::getInt32Ty(LLVMContext));
+    Pad = Handler->CleanupPad =
+        LLVMBuilder->CreateCleanupPad({Undef}, "FaultPad");
+    LLVMBuilder->CreateBr(HandlerCodeBlock);
+    // The in-flight exception is not available to a fault handler.
+    Exception = nullptr;
+    break;
+  }
+  case ReaderBaseNS::RegionKind::RGN_MExcept: {
+    Type *ExceptionType =
+        getType(CORINFO_TYPE_CLASS,
+                getBuiltinClass(CorInfoClassId::CLASSID_SYSTEM_OBJECT));
+    // The filter returns an i32 (though its value is always either 0 or 1)
+    Type *Int8PtrTy = Type::getInt8PtrTy(LLVMContext);
+    Type *Int32Ty = Type::getInt32Ty(LLVMContext);
+    FunctionType *FilterTy =
+        FunctionType::get(Int32Ty, {Int8PtrTy, ExceptionType}, false);
+
+    // Create the filter function.
+    uint32_t FilterOffset = FilterRegion->StartMsilOffset;
+    Twine FunctionName = Twine("Filter_") + Twine::utohexstr(FilterOffset);
+    Function *FilterFunction =
+        Function::Create(FilterTy, Function::ExternalLinkage, FunctionName,
+                         JitContext->CurrentModule);
+    FilterFunction->addFnAttr("wineh-parent", RootFunction->getName());
+    assert(PersonalityFunction != nullptr);
+    FilterFunction->setPersonalityFn(PersonalityFunction);
+    FilterFunction->setGC("coreclr");
```

Comment:
Nit: it might be better to write this as `FilterFunction->setGC(RootFunction->getGC())` or at least assert that the GC of the funclet matches that of its parent.